src/v/kafka/server/handlers/describe_configs.h

```diff
@@ -9,12 +9,84 @@
  * by the Apache License, Version 2.0
  */
 #pragma once
+#include "config/configuration.h"
+#include "config/data_directory_path.h"
+#include "config/node_config.h"
 #include "kafka/protocol/describe_configs.h"
 #include "kafka/server/handlers/handler.h"
+#include "kafka/server/handlers/topics/types.h"
+#include "model/fundamental.h"
+#include "reflection/type_traits.h"
 
 namespace kafka {
 
 using describe_configs_handler
   = single_stage_handler<describe_configs_api, 0, 4>;
 
+// property_config_type maps the datatype for a config property to
+// describe_configs_type. Currently class_type and password are not used in
+// Redpanda so we do not include checks for those types. You may find a similar
+// mapping in Apache Kafka at
+// https://github.com/apache/kafka/blob/be032735b39360df1a6de1a7feea8b4336e5bcc0/core/src/main/scala/kafka/server/ConfigHelper.scala
+// Type sizes are based on Java and taken from
+// https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
+template<typename T>
+describe_configs_type property_config_type(std::string_view name) {
```

Comment:
Figuring out how to call `property_config_type` in the unit test while keeping this method in the .cc file was taking to much time and  I kept getting an `undefined symbol` linker error. Any suggestions on ways to keep this method in an implementation file so we can reduce compile times?

An example linker error when I had the method in the anon namespace:
```
FAILED: bin/test_kafka_server_fixture_rpunit 
: && /home/nyalia/redpanda/vbuild/llvm/install/bin/clang++ -fPIC -march=westmere -fuse-ld=lld -mllvm -inline-threshold=2500 -Wno-unused-lambda-capture -Wno-unused-command-line-argument -stdlib=libc++ -fsanitize=undefined -fsanitize=address -O0 -g -gdwarf-4 -gsplit-dwarf -Wl,--build-id=sha1 -L/home/nyalia/redpanda/vbuild/debug/clang/rp_deps_install/lib -L/home/nyalia/redpanda/vbuild/debug/clang/rp_deps_install/lib64 -stdlib=libc++ -fsanitize=undefined -fsanitize=address -pie -lc++experimental   -rdynamic src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/consumer_groups_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/token_bucket_algorithm_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/member_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/group_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/read_write_roundtrip_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/fetch_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/delete_topics_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/offset_fetch_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/api_versions_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/create_topics_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/find_coordinator_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/list_offsets_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/topic_recreate_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/fetch_session_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/alter_config_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/produce_consume_test.cc.o src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/group_metadata_serialization_test.cc.o -o bin/test_kafka_server_fixture_rpunit  -Wl,-rpath,/home/nyalia/redpanda/vbuild/debug/clang/rp_deps_install/lib:  lib/libv_v_seastar_testing_main.a  lib/libv_v_application.a  lib/libv_v_raft.a  lib/libv_v_kafka.a  lib/libv_v_config.a  lib/libv_v_storage_test_utils.a  rp_deps_install/lib/libseastar_testing.a  rp_deps_install/lib/libboost_unit_test_framework.so  lib/libv_v_pandaproxy_rest.a  lib/libv_v_pandaproxy_schema_registry.a  lib/libv_v_pandaproxy_common.a  lib/libv_v_kafka.a  lib/libv_v_coproc.a  lib/libv_v_kafka_client.a  lib/libv_v_kafka.a  lib/libv_v_coproc.a  lib/libv_v_kafka_client.a  lib/libv_v_kafka_protocol.a  lib/libcoproc_gen.a  lib/libv_v_pandaproxy_parsing.a  lib/libv_v_pandaproxy_json.a  -lavrocpp_s  rp_deps_install/lib/libboost_iostreams.so  rp_deps_install/lib/libboost_regex.so  rp_deps_install/lib/libprotocd.a  rp_deps_install/lib/libprotobufd.a  rp_deps_install/lib/libz.so  lib/libv_v_archival.a  lib/libv_v_cluster.a  lib/libv_v_raft.a  lib/libraft_rpc.a  lib/libbootstrap_rpc.a  lib/libcontroller_rpc.a  lib/libmetadata_rpc.a  lib/libid_allocator_rpc.a  lib/libtx_gateway_rpc.a  lib/libpartition_balancer_rpc.a  lib/libnode_status_rpc.a  lib/libephemeral_credential_rpc.a  lib/libv_v_rpc.a  rp_deps_install/lib/libroaring.a  lib/libv_v_v8_engine.a  lib/libv_v_features.a  lib/libv_v_cloud_storage.a  lib/libv_v_s3.a  lib/libv_v_cloud_roles.a  lib/libv_v_http.a  lib/libv_v_net.a  lib/libv_v_storage.a  lib/libv_v_config.a  lib/libv_v_security.a  -lcryptopp  lib/libv_v_json.a  -lboost_filesystem  lib/libv_v_syschecks.a  lib/libv_v_utils.a  rp_deps_install/lib/libhdr_histogram_static.a  rp_deps_install/lib/libbase64.a  lib/libv_v_finjector.a  rp_deps_install/lib/libabsl_hash.a  rp_deps_install/lib/libabsl_city.a  rp_deps_install/lib/libabsl_bad_variant_access.a  rp_deps_install/lib/libabsl_low_level_hash.a  rp_deps_install/lib/libabsl_raw_hash_set.a  rp_deps_install/lib/libabsl_hashtablez_sampler.a  lib/libv_v_resource_mgmt.a  rp_deps_install/lib/libabsl_cord.a  rp_deps_install/lib/libabsl_cordz_info.a  rp_deps_install/lib/libabsl_cord_internal.a  rp_deps_install/lib/libabsl_cordz_functions.a  rp_deps_install/lib/libabsl_exponential_biased.a  rp_deps_install/lib/libabsl_cordz_handle.a  rp_deps_install/lib/libabsl_synchronization.a  rp_deps_install/lib/libabsl_graphcycles_internal.a  rp_deps_install/lib/libabsl_symbolize.a  rp_deps_install/lib/libabsl_malloc_internal.a  rp_deps_install/lib/libabsl_demangle_internal.a  rp_deps_install/lib/libabsl_time.a  rp_deps_install/lib/libabsl_civil_time.a  rp_deps_install/lib/libabsl_time_zone.a  rp_deps_install/lib/libabsl_stacktrace.a  rp_deps_install/lib/libabsl_debugging_internal.a  rp_deps_install/lib/libroaring.a  lib/libv_v_model_test_utils.a  lib/libv_v_model.a  lib/libv_v_rphashing.a  rp_deps_install/lib/libxxhash.a  rp_deps_install/lib/libcrc32c.a  lib/libv_v_compression.a  rp_deps_install/lib/libzstd.a  rp_deps_install/lib/liblz4.so  rp_deps_install/lib/libsnappy.a  rp_deps_install/lib/libz.so  lib/libv_v_rprandom.a  lib/libv_v_bytes.a  rp_deps_install/lib/libseastar.a  rp_deps_install/lib/libboost_program_options.so  rp_deps_install/lib/libboost_thread.so  rp_deps_install/lib/libboost_chrono.so  rp_deps_install/lib/libboost_date_time.so  rp_deps_install/lib/libboost_atomic.so  rp_deps_install/lib/libcares.so  rp_deps_install/lib/libcryptopp.so  rp_deps_install/lib/libfmtd.a  rp_deps_install/lib/liblz4.so  -ldl  rp_deps_install/lib/libgnutls.so  -latomic  rp_deps_install/lib/libsctp.so  rp_deps_install/lib/libyaml-cpp.a  -fsanitize=address  -fsanitize=undefined  -fno-sanitize=vptr  rp_deps_install/lib/libhwloc.so  rp_deps_install/lib/liburing.so  -Wl,--build-id=sha1 -L/home/nyalia/redpanda/vbuild/debug/clang/rp_deps_install/lib -L/home/nyalia/redpanda/vbuild/debug/clang/rp_deps_install/lib64 -stdlib=libc++ -fsanitize=undefined -fsanitize=address -pthread -lc++experimental  rp_deps_install/lib/libnuma.so  rp_deps_install/lib/libabsl_random_seed_sequences.a  rp_deps_install/lib/libabsl_random_internal_pool_urbg.a  rp_deps_install/lib/libabsl_random_internal_randen.a  rp_deps_install/lib/libabsl_random_internal_randen_hwaes.a  rp_deps_install/lib/libabsl_random_internal_randen_hwaes_impl.a  rp_deps_install/lib/libabsl_random_internal_randen_slow.a  rp_deps_install/lib/libabsl_random_internal_platform.a  rp_deps_install/lib/libabsl_random_internal_seed_material.a  rp_deps_install/lib/libabsl_strings.a  rp_deps_install/lib/libabsl_strings_internal.a  rp_deps_install/lib/libabsl_base.a  rp_deps_install/lib/libabsl_spinlock_wait.a  -lrt  rp_deps_install/lib/libabsl_int128.a  rp_deps_install/lib/libabsl_bad_optional_access.a  rp_deps_install/lib/libabsl_random_seed_gen_exception.a  rp_deps_install/lib/libabsl_throw_delegate.a  rp_deps_install/lib/libabsl_raw_logging_internal.a  rp_deps_install/lib/libabsl_log_severity.a && :
ld.lld: error: undefined symbol: kafka::(anonymous namespace)::property_config_type<config::one_or_many_property<config::broker_authn_endpoint> >::get(std::__1::basic_string_view<char, std::__1::char_traits<char> >)
>>> referenced by alter_config_test.cc:60 (/home/nyalia/redpanda/src/v/kafka/server/tests/alter_config_test.cc:60)
>>>               src/v/kafka/server/tests/CMakeFiles/test_kafka_server_fixture_rpunit.dir/alter_config_test.cc.o:(alter_config_test_fixture::config_type_and_documentation alter_config_test_fixture::make_config_type_and_documentation<config::one_or_many_property<config::broker_authn_endpoint> >(config::one_or_many_property<config::broker_authn_endpoint> const&, std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::basic_string_view<char, std::__1::char_traits<char> >))
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
[56/56] Linking CXX executable bin/redpanda
```

Reply:
I think its ok to not test this explicitly since its a `consteval` method its outputs are what you'd see in the source at compile time. 

You can implicitly test this by writing ducktape unit tests that test for a property that contains a config type that you expect to be `int` or `short` or `list` ... etc

Reply:
I got to create the ducktape test in the next push. I'm currently messing around with kafka-python's client but their library does not have a DescribeConfigsRequest so I'm reading through the docs to learn how to create my own.

Reply:
I'm going to look into KCL instead